core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java

```diff
@@ -161,33 +154,38 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         String groupBy = params.param("group_by", "nodes");
         if ("nodes".equals(groupBy)) {
+            if (discoveryNodes == null) {
+                throw new IllegalStateException("discoveryNodes must be set before calling toXContent with group_by=nodes");
+            }
             builder.startObject("nodes");
-            for (Map.Entry<DiscoveryNode, List<TaskInfo>> entry : getPerNodeTasks().entrySet()) {
-                DiscoveryNode node = entry.getKey();
-                builder.startObject(node.getId());
-                builder.field("name", node.getName());
-                builder.field("transport_address", node.getAddress().toString());
-                builder.field("host", node.getHostName());
-                builder.field("ip", node.getAddress());
-
-                builder.startArray("roles");
-                for (DiscoveryNode.Role role : node.getRoles()) {
-                    builder.value(role.getRoleName());
-                }
-                builder.endArray();
-
-                if (!node.getAttributes().isEmpty()) {
-                    builder.startObject("attributes");
-                    for (Map.Entry<String, String> attrEntry : node.getAttributes().entrySet()) {
-                        builder.field(attrEntry.getKey(), attrEntry.getValue());
+            for (Map.Entry<String, List<TaskInfo>> entry : getPerNodeTasks().entrySet()) {
+                DiscoveryNode node = discoveryNodes.get(entry.getKey());
+                builder.startObject(entry.getKey());
+                if (node != null) {
+                    // If the node is no longer part of the cluster, oh well, we'll just skip it's useful information.
+                    builder.field("name", node.getName());
+                    builder.field("transport_address", node.getAddress().toString());
+                    builder.field("host", node.getHostName());
+                    builder.field("ip", node.getAddress());
+
+                    builder.startArray("roles");
+                    for (DiscoveryNode.Role role : node.getRoles()) {
+                        builder.value(role.getRoleName());
+                    }
+                    builder.endArray();
+
+                    if (!node.getAttributes().isEmpty()) {
+                        builder.startObject("attributes");
+                        for (Map.Entry<String, String> attrEntry : node.getAttributes().entrySet()) {
+                            builder.field(attrEntry.getKey(), attrEntry.getValue());
+                        }
+                        builder.endObject();
                     }
-                    builder.endObject();
                 }
                 builder.startObject("tasks");
                 for(TaskInfo task : entry.getValue()) {
-                    builder.startObject(task.getTaskId().toString());
+                    builder.field(task.getTaskId().toString());
```

Comment:
This kind of got pulled along. ToXContent doesn't specify either way but we have lots of stuff that acts like it does spit out a whole object. `Strings#toString`, for example.

Anyway, I hit some broken thing with XContent and started pulling on it to "fix" it the way I was used to seeing other things and it sort of expanded into a lot of changes. I can revert it and break it and deal with it later if you like.

Part of the reason I like it this was is that it means that ToXContent has symmetry with the parser. You can parse exactly what it spits out with what it came from.